electrumsv/network_support/esv_client.py

```diff
@@ -0,0 +1,415 @@
+
+            result = []
+            for json_token in json_tokens:
+                permissions = TokenPermissions.NONE
+                if json_token['can_read']:
+                    permissions |= TokenPermissions.READ_ACCESS
+                if json_token['can_write']:
+                    permissions |= TokenPermissions.WRITE_ACCESS
+                result.append(PeerChannelToken(permissions=permissions,
+                    api_key=json_token['token']))
+            return result
+
+
+class ESVClient:
+    """This is a lightweight client for the ElectrumSVReferenceServer.
+
+    The only state is the base_url and master_token. Therefore instances of ESVClient can be
+    re-generated on-demand - no need for caching of ESVClient instances."""
+
+    def __init__(self, base_url: str, session: aiohttp.ClientSession, master_token: str):
+        self.base_url = base_url
+        self.session = session
+        self.master_token = master_token
+        self.headers = {"Authorization": f"Bearer {self.master_token}"}
+
+        self._message_fetcher_is_alive = False
+        self._FETCH_JOBS_COUNT = 4
+        self._merkle_proofs_queue: asyncio.Queue[MAPICallbackResponse] = asyncio.Queue()
+        self._peer_channel_cache: Dict[ChannelId, PeerChannel] = {}
+
+    def _replace_http_with_ws(self, url: str) -> str:
+        if url.startswith("http://"):
+            url = self.base_url.replace("http://", "ws://")
+        if self.base_url.startswith("https://"):
+            url = url.replace("https://", "wss://")
+        return url
+
+    def _peer_channel_json_to_obj(self, peer_channel_json: PeerChannelViewModelGet) \
+            -> PeerChannel:
+        access_tokens = peer_channel_json['access_tokens']
+        tokens = []
+        for token in access_tokens:
+            permissions = TokenPermissions.NONE
+            if token['can_read']:
+                permissions |= TokenPermissions.READ_ACCESS
+            if token['can_write']:
+                permissions |= TokenPermissions.WRITE_ACCESS
+            tokens.append(PeerChannelToken(permissions=permissions, api_key=token['token']))
+
+        return PeerChannel(channel_id=peer_channel_json['id'], tokens=tokens,
+            base_url=self.base_url, session=self.session, master_token=self.master_token)
+
+    # ----- General Websocket ----- #
+    async def _fetch_peer_channel_message_job(self,
+            peer_channel_notification_queue: asyncio.Queue[GeneralNotification]) -> None:
+        """Can run multiple of these concurrently to fetch new peer channel messages"""
+        message: GeneralNotification = await peer_channel_notification_queue.get()
+        channel_id: ChannelId = message['result']['id']
+
+        peer_channel = await self.get_single_peer_channel_cached(channel_id)
+        messages: list[PeerChannelMessage] = await peer_channel.get_messages()  # network io
+        for pc_message in messages:
+            if pc_message['content_type'] == 'application/json':
```

Comment:
Get rid of the json support, never do it unless we are forced to by some protocol.

Reply:
See my earlier comment. mAPI always gives us json payloads so I am not sure if this is viable.